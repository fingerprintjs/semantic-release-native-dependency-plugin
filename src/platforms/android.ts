import { humanizeMavenStyleVersionRange } from '../utils'
import type { GenerateNotesContext } from 'semantic-release'
import type { Signale } from 'signale'
import { join } from 'node:path'
import { getCommand } from '../gradle'
import { spawn } from 'node:child_process'

export interface AndroidPlatformConfiguration {
  path: string | undefined
  gradleTaskName: string | undefined
  displayName: string | undefined
}

async function runGradleTask(
  cwd: string,
  androidPath: string,
  androidGradleTaskName: string,
  env: NodeJS.ProcessEnv,
  logger: Signale
): Promise<string> {
  const androidFullPath = join(cwd, androidPath)
  const command = await getCommand(androidFullPath)

  return new Promise((resolve, reject) => {
    const child = spawn(command, [androidGradleTaskName, '-q', '--console=plain'], {
      cwd: androidFullPath,
      env,
      detached: true,
      stdio: ['inherit', 'pipe', 'pipe'],
    })
    if (child.stdout === null) {
      reject(new Error('Unexpected error: stdout of subprocess is null'))
      return
    }
    if (child.stderr === null) {
      reject(new Error('Unexpected error: stderr of subprocess is null'))
      return
    }

    let androidVersion: string | null = null
    child.stdout.on('data', (line: Buffer) => {
      if (!line || line.toString().trim() === '') {
        return
      }

      logger.debug(`Gradle stdout: ${line}`)
      androidVersion = line.toString().trim()
    })
    child.stderr.on('data', (line: Buffer) => {
      logger.error(line.toString().trim())
    })
    child.on('close', (code: number) => {
      if (code !== 0) {
        reject(new Error(`Unexpected error: Gradle failed with status code ${code}`))
        return
      }

      if (androidVersion === null) {
        reject(new Error(`Could not read output of \`${androidGradleTaskName}\` gradle task.`))
        return
      }

      resolve(androidVersion)
    })
    child.on('error', (err) => {
      logger.error(err)
      reject(err)
    })
  })
}

export type AndroidResolveContext = Pick<GenerateNotesContext, 'logger' | 'cwd' | 'env'>

export const resolve = async (
  { logger, cwd, env }: AndroidResolveContext,
  { path, gradleTaskName }: AndroidPlatformConfiguration
) => {
  if (!cwd) {
    throw new Error(`Current working directory is required to detect android dependency version range.`)
  }

  if (!path) {
    throw new Error('Android path should be defined.')
  }

  if (!gradleTaskName) {
    throw new Error('Android gradle task name should be defined.')
  }

  const androidVersion = await runGradleTask(cwd, path, gradleTaskName, env as NodeJS.ProcessEnv, logger)
  return humanizeMavenStyleVersionRange(androidVersion)
}
