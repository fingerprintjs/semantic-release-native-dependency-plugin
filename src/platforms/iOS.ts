import { join } from 'node:path'
import { readFileSync } from 'node:fs'
import type { GenerateNotesContext } from 'semantic-release'

export interface IOSPlatformConfiguration {
  podSpecJsonPath: string | undefined
  dependencyName: string | undefined
  displayName: string | undefined
}

type PodspecJson = {
  dependencies: {
    [key: string]: [string]
  }
}

export type IOSResolveContext = Pick<GenerateNotesContext, 'cwd'>

export const resolve = async (
  { cwd }: IOSResolveContext,
  { podSpecJsonPath, dependencyName }: IOSPlatformConfiguration
) => {
  if (!cwd) {
    throw new Error(`Current working directory is required to detect iOS dependency version range.`)
  }

  if (!podSpecJsonPath) {
    throw new Error('iOS Podspec Json path should be defined.')
  }

  if (!dependencyName) {
    throw new Error('iOS Dependency name should be defined.')
  }

  const jsonFile = join(cwd, podSpecJsonPath)

  let fileContent: string
  try {
    fileContent = readFileSync(jsonFile, 'utf8')
  } catch (error: any) {
    switch (error.code) {
      case 'ENOENT':
        throw new Error(`${podSpecJsonPath} file does not exist.`)
      case 'EACCES':
        throw new Error(`${podSpecJsonPath} file cannot be accessed.`)
      default:
        throw new Error(`${podSpecJsonPath} file cannot be read. Error: ${error.message}`)
    }
  }

  let data: PodspecJson
  try {
    data = JSON.parse(fileContent) as PodspecJson
  } catch (error) {
    throw new Error(`${podSpecJsonPath} file cannot be parsed as JSON.`)
  }

  if (!data.dependencies || !data.dependencies[dependencyName]) {
    throw new Error(`${podSpecJsonPath} file does not contain '${dependencyName}' in dependencies.`)
  }

  return data.dependencies[dependencyName].join(' and ')
}
