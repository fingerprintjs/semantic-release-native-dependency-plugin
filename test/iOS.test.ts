import { describe, expect, it } from '@jest/globals'
import { cwd } from 'node:process'
import { IOSResolveContext, resolve } from '../src/platforms/iOS'

describe('Test for iOS platform', () => {
  const dependencyName = 'FingerprintPro'
  const ctx = {
    cwd: cwd(),
  } satisfies IOSResolveContext

  it('returns correct version', async () => {
    const iosVersion = await resolve(ctx, {
      podSpecJsonPath: 'test/project/ios/podspec.json',
      dependencyName,
      displayName: undefined,
    })
    expect(iosVersion).toBe('>= 1.2.3 and < 4.5.6')
  })

  it('throws error when file not found', async () => {
    await expect(
      resolve(ctx, {
        podSpecJsonPath: 'nonExists',
        dependencyName,
        displayName: undefined,
      })
    ).rejects.toThrowErrorMatchingSnapshot('podspecFileNotFound')
  })

  it('throws error when file not readable', async () => {
    const readFileSyncMock = jest.spyOn(require('node:fs'), 'readFileSync')
    readFileSyncMock.mockImplementation(() => {
      throw { code: 'EACCES' }
    })

    await expect(
      resolve(ctx, { podSpecJsonPath: 'not-readable.podspec.json', dependencyName, displayName: undefined })
    ).rejects.toThrowErrorMatchingSnapshot('podspecNotReadable')

    readFileSyncMock.mockRestore()
  })

  it('throws error when path is a directory', async () => {
    await expect(
      resolve(ctx, {
        podSpecJsonPath: 'test',
        dependencyName,
        displayName: undefined,
      })
    ).rejects.toThrowErrorMatchingSnapshot('podspecPathIsDirectory')
  })
})
