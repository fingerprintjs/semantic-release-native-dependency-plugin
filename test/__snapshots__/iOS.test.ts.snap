// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for iOS platform throws error when file not found: podspecFileNotFound 1`] = `"nonExists file does not exist."`;

exports[`Test for iOS platform throws error when file not readable: podspecNotReadable 1`] = `"not-readable.podspec.json file cannot be accessed."`;

exports[`Test for iOS platform throws error when path is a directory: podspecPathIsDirectory 1`] = `"test file cannot be read. Error: EISDIR: illegal operation on a directory, read"`;
